### ISSUE TITLE
Player login packet header sometimes becomes garbled

### ISSUE THREAD

**Issue by Kangaroux** [Kangaroux/gomaggus#18]
Sometimes the login packet header is garbled, and it consistently happens when receiving the login packet. Once this happens, all future incoming headers are garbled. I'm not sure which side is causing it, but the RC4 stream is getting out of sync and makes the connection useless.

I suspect the reason it only happens sometimes is because the cipher is seeded using the session key. This could be verified by replaying the packets with the same session key and seeing if it consistently breaks. If it's reproducible based on the session key, there may be a pattern that can be identified by generating random keys and seeing which break. For example, a leading or trailing zero in the key. 

I confirmed using wireshark that there aren't any packets the server is missing. The garbled packet is definitely for the player login. The payload body is 14 bytes (6+8) and the body is the 64 bit char ID. The client also consistently sends this after realm split.

```
2024/06/18 18:59:05 RECV  op=0x38c   dsize=92     hsize=8
2024/06/18 18:59:05 SENT  op=0x38b   size=17
2024/06/18 18:59:05 sent realm split
2024/06/18 18:59:09 RECV  op=0x6e5b92f8  dsize=14     hsize=1396
2024/06/18 18:59:09 unknown opcode: 0x6e5b92f8
```

**Comment by Kangaroux** [comment]
After a bunch of manual testing I had this issue happen as the client was authenticating to realmd. The header encryption is NOT enabled, so it wasn't a cipher issue.

I started logging the contents of every packet that came in and compared them with wireshark. I noticed that I sometimes saw a large packet being logged but in wireshark no such packet existed.

> If you've worked with TCP before you know what's coming

```
76 bytes
577010d0e0f1e41f0000cd774de6580c5f240000bcc2cd1d70ebea46000090581f03dea594520000a72fcd5583132a560000d83759d4f75c7b8500002e8827bf2f4deeb60000cdc88fb817e6
```

The packet I was expecting to see was this:

![image](https://github.com/Kangaroux/gomaggus/assets/2302541/f47ae540-5894-4d01-bbef-e0a2522629e0)

The logged payload does contain the correct bytes but the read count reported by `Read()` is wrong. I looked at the next packet in wireshark:

![image](https://github.com/Kangaroux/gomaggus/assets/2302541/7f701bb7-e326-4543-9ea1-033e756fddfd)

Sure enough, that packet is in the log, too. 

This lead me down a rabbit hole, reading every stack overflow or blog post I could find. Eventually I found a great post by [James Stanley](https://incoherency.co.uk/blog/stories/reading-tcp-sockets.html) that describes exactly my issue.

tl;dr: There is no guarantee that calling `Read()` on a TCP socket will give you a _packet_. TCP may decide to fragment a packet into smaller pieces or combine several small packets into a larger one. The only guarantee TCP provides when reading is that the bytes will be in order. It might be part of a packet, exactly one packet, many packets, etc.
